frontend/model/notifications/mainNotificationsMixin.js

```diff
@@ -127,13 +131,60 @@ const periodicNotificationEntries = [
         return comparePeriodStamps(new Date().toISOString(), rootGetters.groupSettings.distributionDate) > 0
       }
     }
+  },
+  {
+    type: PERIODIC_NOTIFICATION_TYPE.MIN15,
+    notificationData: {
+      stateKey: 'expiringProposals',
+      emitCondition ({ rootGetters }) {
+        this.expiringProposalsByGroup = rootGetters.groupsByName.map(group => {
+          const { contractID } = group
+          const expiringProposals = []
+          const groupProposals = rootGetters.groupProposals(contractID) || {}
+          for (const proposalId in groupProposals) {
+            const proposal = groupProposals[proposalId]
+            if (proposal.status === STATUS_OPEN &&
+              !proposal.notifiedBeforeExpire &&
+              proposal.data.expires_date_ms < (Date.now() + DAYS_MILLIS)) {
+              expiringProposals.push({
+                proposalId,
+                proposalType: proposal.data.proposalType,
+                proposalData: proposal.data.proposalData,
+                expires_date_ms: proposal.data.expires_date_ms,
+                createdDate: proposal.meta.createdDate,
+                creator: proposal.meta.username
+              })
+            }
+          }
+
+          return { contractID, proposals: expiringProposals }
+        }).filter(entry => entry.proposals.length)
+
+        return this.expiringProposalsByGroup.length
+      },
+      async emit () {
+        for (const { contractID, proposals } of this.expiringProposalsByGroup) {
+          await sbp('gi.actions/group/notifyExpiringProposals', {
+            contractID,
+            data: { proposals }
+          })
```

Comment:
So the idea is ultimately to notify the user of something without touching the contract code.... and I'm trying to figure out what this code does and where the notification is ultimately sent. And I'm coming to the conclusion that the notification itself was never implemented!

Because this action sends a message to the contracts that ultimately just sets a `notifiedBeforeExpire` variable on a particular proposal to indicate that we've notified the user about this about-to-expire proposal - but I can't find where that notification is actually sent. I don't think it's sent?

So - we do need a way to keep track of whether or not we've notified about a particular proposal. So for that we can use the `state` that you've created, and use the `proposalId` to key an object on the state as to whether or not we've already emitted a notification for that proposal, e.g. `this.proposalNotificationSent[proposalId]`, and use that instead of `notifiedBeforeExpire`

Reply:
One more thought - regarding the design of using `this.proposalNotificationSent[proposalId]` to keep track of notified expiring proposals - as far as I can tell, that state will not be cleared until the browser page is refreshed.

And in this case `shouldClearStateKey` isn't being used for this purpose, because this notification really has within it a whole bunch of other notifications (one for each expiring proposal).

Just something to be aware of, maybe deserving of a comment too, because it's different from how the other notifications behave.

Reply:
@taoeffect 
> but I can't find where that notification is actually sent. I don't think it's sent?
Yes, it looks like the 'gi.contracts/group/notifyExpiringProposals' only toggles `notifiedBeforeExpire` prop in the proposal object but doesn't actually send the notification as the action side-effect. So it seems like a new notification type (e.g. PROPOSAL_EXPIRING) has to be added and the logic for sending it needs to be implemented too.

`this.proposalNotificationSent[proposalId]` sounds like a good idea.
Btw, there is another way of checking if a particular notification has already been sent for the active user that I've created while working on this PR, which is `myNotificationHas` method (L9 in `mainNotificationsMixin.js`). we can use this method as well too.

Lastly, this is a topic that I have never looked at so far, so even though this is fairly relevant to the original issue(which is implementing various types of notifications), but quite overwhelming. So is it alright for this additional work to be handled as a separate issue in the future either by me or someone else?

Thanks,

Reply:
> Lastly, this is a topic that I have never looked at so far, so even though this is fairly relevant to the original issue(which is implementing various types of notifications), but quite overwhelming. So is it alright for this additional work to be handled as a separate issue in the future either by me or someone else?

Yep! 👍

Reply:
Once, the issue is created, please assign myself to it.